component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/voiceotp/VoiceOTPAuthenticator.java

```diff
@@ -0,0 +1,2037 @@
+            throws AuthenticationFailedException, VoiceOTPException {
+        //the authentication flow happens with voice otp authentication.
+        String tenantDomain = context.getTenantDomain();
+        String errorPage = getErrorPage(context);
+        if (context.isRetrying() && !Boolean.parseBoolean(request.getParameter(VoiceOTPConstants.RESEND))
+                && !isMobileNumberUpdateFailed(context)) {
+
+            log.debug("Trigger retry flow when it is not " +
+                        "request for resending OTP or it is not mobile number update failure");
+
+            checkStatusCode(response, context, queryParams, errorPage);
+        } else {
+            processVoiceOTPFlow(context, request, response, isUserExists, username, queryParams, tenantDomain,
+                    errorPage);
+        }
+    }
+
+    /**
+     *
+     * @param context The AuthenticationContext
+     * @param response The HttpServletResponse
+     * @param username username
+     * @param queryParams request query params
+     * @param sendOtpToFederatedMobile whether it needs to send the otp for federated user mobile number
+     * @throws AuthenticationFailedException
+     */
+    private void proceedOTPWithFederatedMobileNumber(AuthenticationContext context, HttpServletResponse response,
+                                                     String username, String queryParams,
+                                                     boolean sendOtpToFederatedMobile)
+            throws AuthenticationFailedException, VoiceOTPException {
+
+        try {
+            String federatedMobileAttributeKey;
+            String mobile = null;
+            StepConfig stepConfig = context.getSequenceConfig().getStepMap().get(context.getCurrentStep() - 1);
+            String previousStepAuthenticator = stepConfig.getAuthenticatedAutenticator().getName();
+            StepConfig currentStep = context.getSequenceConfig().getStepMap().get(context.getCurrentStep());
+            String currentStepAuthenticator = currentStep.getAuthenticatorList().iterator().next().getName();
+            if (sendOtpToFederatedMobile) {
+                federatedMobileAttributeKey = getFederatedMobileAttributeKey(context, previousStepAuthenticator);
+                if (StringUtils.isEmpty(federatedMobileAttributeKey)) {
+                    federatedMobileAttributeKey = getFederatedMobileAttributeKey(context, currentStepAuthenticator);
+                }
+                Map<ClaimMapping, String> userAttributes = context.getCurrentAuthenticatedIdPs().values().
+                        iterator().next().getUser().getUserAttributes();
+                for (Map.Entry<ClaimMapping, String> entry : userAttributes.entrySet()) {
+                    String key = String.valueOf(entry.getKey().getLocalClaim().getClaimUri());
+                    String value = entry.getValue();
+                    if (key.equals(federatedMobileAttributeKey)) {
```

Comment:
Shall we add a null check here?
```suggestion
                    if (key != null && key.equals(federatedMobileAttributeKey)) {
```